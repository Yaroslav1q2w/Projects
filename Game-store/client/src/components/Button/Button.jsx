import React from 'react';
import "./Button.scss"
import PropTypes from "prop-types";

const Button = ({children, onClick,className,type}) => {

    return (
        <button type={type} className={className} onClick={onClick}>{children}</button>
    );
};


// Button.propTypes = {
//     children: PropTypes.string.isRequired,
//     onClick: PropTypes.func.isRequired,
// }

export default Button;
