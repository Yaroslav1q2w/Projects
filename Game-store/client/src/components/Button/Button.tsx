interface IProps {
	children: React.ReactNode;
	onClick: () => void;
	className: string;
	type: "submit" | "reset" | "button";
}

const Button = ({ children, onClick, className, type }: IProps) => {
	return (
		<button type={type} className={className} onClick={onClick}>
			{children}
		</button>
	);
};

export default Button;
