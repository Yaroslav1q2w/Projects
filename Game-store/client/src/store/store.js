import { configureStore } from "@reduxjs/toolkit";

import {
	favoriteReducer,
	todoReducer,
	basketReducer,
	modalReducer,
	categoriesReducer,
} from "../reducers";

const store = configureStore({
	reducer: {
		favorite: favoriteReducer,
		todos: todoReducer,
		basket: basketReducer,
		modal: modalReducer,
		category: categoriesReducer,
	},
});

export default store;
