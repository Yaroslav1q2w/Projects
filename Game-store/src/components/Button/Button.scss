
.button {
    font-family: 'Montserrat', sans-serif;
    font-style: normal;
    width: 140px;
    height: 35px;
    color: #fafafa;
    //background: linear-gradient(to left, #16162d, #012344,#16162d);
    background: #26264f;
    border: 0;
    border-radius: 10px;
    outline: none;
    margin: 6px;
    cursor: pointer;
    transition: .5s;

    &:hover {
        background: #323275;
    }
}

.button__modals {
    display: flex;
    justify-content: center;
    align-items: center;
    padding-bottom: 20px;
}
