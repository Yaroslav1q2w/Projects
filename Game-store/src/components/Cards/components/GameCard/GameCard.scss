.game__item {
    color: white;
    width: 320px;
    min-height: 300px;
    background-color: rgb(36,37,41);
    border-radius: 15px;
    position: relative;
    margin-bottom: 30px;
    justify-self: center;
    
    &-details {
        padding: 10px 20px;
        position: relative;
    }
    
    &-header {
        display: flex;
        justify-content: space-between;
        align-items: baseline;
    }
    
    &-title {
        padding-top: 10px;
        font-size: 19px;
        font-weight: bold;
        margin: 0;
    }
    
    .game__item-article {
        display: inline-block;
        padding-top: 8px;
        font-size: 12px;
        font-weight: 200;
        color: #7a7a7a;
    }
    
    &-description {
        display: flex;
        justify-content: space-between;
        align-items: center;
        margin-top: 10px;
        font-size: 14px;
        position: relative;
    }
    
    .game__item-price {
        font-size: 16px;
    }
}

.star-icon {
    position: absolute;
    top: 17px;
    right: 7px;
    cursor: pointer;
}