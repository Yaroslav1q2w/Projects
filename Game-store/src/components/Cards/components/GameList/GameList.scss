.list__item {
    color: white;
    width: 100%;
    position: relative;
    margin-top: 20px;
    padding-top: 20px;
    justify-self: center;
    display: flex;
    border-top: 1px solid #343434;
    
    &-details {
        padding: 0 20px;
        position: relative;
        width: 100%;
    }
    
    &-favorite {
        position: absolute;
        bottom: 0;
        cursor: pointer;
    }
    
    &-header {
        display: flex;
        flex-direction: column;
        align-items: start;
    }
    
    &-title {
        padding-top: 10px;
        font-size: 20px;
        font-weight: bold;
        margin: 0;
    }
    
    &-article {
        display: inline-block;
        padding-top: 8px;
        font-size: 12px;
        font-weight: 200;
        color: #7a7a7a;
    }
    
    &-description {
        display: flex;
        flex-direction: column;
        justify-content: space-between;
        align-items: center;
        font-size: 14px;
        position: relative;
        
    }
    &-price {
        font-size: 18px;
    }
}
