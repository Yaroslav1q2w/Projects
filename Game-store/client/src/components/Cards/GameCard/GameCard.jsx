import React,{useState} from 'react';
import Button from "../../Button";
import "./GameCard.scss"
import {AiFillStar, AiOutlineStar} from "react-icons/ai";
import {useDispatch } from 'react-redux'
import {increaseFavorite,decreaseFavorite,} from "../../../reducers"
import {Link} from "react-router-dom";
import 'swiper/css';


const GameCard = ({ cardProps, isOpenModal, addToCard}) => {
    const {title, article, price, image, description, genre, _id} = cardProps

    const favoritesLocalStorage = JSON.parse(localStorage.getItem("favoriteCount"))
    const isFavorite = Boolean(favoritesLocalStorage?.find(favorite => favorite.article === article))

    const [addFavorites,setAddFavorites] = useState(!isFavorite ? false : true)
    const [notFavorites,setNotFavorites] = useState(isFavorite ? false : true)

    const dispatch = useDispatch()


    return (
        <div className="game__item">
            <div className="game__item-img">
                <img src={image} alt={title} width={320} height={180}/>
                <div className="game__back">
                    <div className="game__back-wrap">
                        <h2 className="game__back-title">{title}</h2>
                        <p className="game__back-desc">{description}</p>
                        <div className="game__back-link">
                            <Link to={`api/products/${_id}`} className="link">перейти</Link>
                        </div>
                    </div>
                </div>
            </div>
            <div className="game__item-details">
                <div className="game__title">
                    <div className="game__item-header">
                        <Link to={`api/products/${_id}`}><h3 className="game__item-title">{title}</h3></Link>
                        <div className="star-icon">
                            {notFavorites && <AiOutlineStar fontSize={26} onClick={() => {
                                dispatch(increaseFavorite(cardProps))
                                setAddFavorites(true)
                                setNotFavorites(false)
                            }}/>}
                            {addFavorites && <AiFillStar fontSize={26} onClick={() => {
                                dispatch(decreaseFavorite(cardProps))
                                setNotFavorites(true)
                                setAddFavorites(false)
                            }}/>}
                        </div>
                    </div>
                    <span className="game__item-article">Артикул: {article}</span>
                </div>
                <div>
                    <span className="game__item-genre">{genre}</span>
                </div>
                <div className="game__item-description">
                    <p className="game__item-price">{price} грн.</p>
                    <Button type="button"
                            children="В корзину"
                            className="button"
                            onClick={() => {isOpenModal(); addToCard()}}
                    />
                </div>
            </div>
        </div>
    );
};

export default GameCard;


