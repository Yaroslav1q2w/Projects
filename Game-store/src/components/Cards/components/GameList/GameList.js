import React,{useState} from 'react';
import {useDispatch} from "react-redux";
import {AiFillStar, AiOutlineStar} from "react-icons/ai";
import {decreaseFavorite, increaseFavorite} from "../../../../reducers";
import Button from "../../../Button";
import {Link} from "react-router-dom";
import "./GameList.scss"


const GameList = ({ cardProps, isOpenModal, addToCard}) => {
    const {title, article, price, image} = cardProps


    const favoritesLocalStorage = JSON.parse(localStorage.getItem("favoriteCount"))
    const isFavorite = Boolean(favoritesLocalStorage?.find(favorite => favorite.article === article))

    const [addFavorites,setAddFavorites] = useState(!isFavorite ? false : true)
    const [notFavorites,setNotFavorites] = useState(isFavorite ? false : true)

    const dispatch = useDispatch()


    return (
        <div className="list__item">
            <img src={image} alt={title} width={260} height={150}/>
            <div className="list__item-details">
                <div className="list__item-header">
                    <Link to={`/card/${article}`}><h3 className="list__item-title">{title}</h3></Link>
                    <div className="list__item-favorite">
                        {notFavorites && <AiOutlineStar fontSize={30} onClick={() => {
                            dispatch(increaseFavorite(cardProps))
                            setAddFavorites(true)
                            setNotFavorites(false)
                        }}/>}
                        {addFavorites && <AiFillStar fontSize={30} onClick={() => {
                            dispatch(decreaseFavorite(cardProps))
                            setNotFavorites(true)
                            setAddFavorites(false)
                        }}/>}
                    </div>
                </div>
                <span className="list__item-article">Артикул: {article}</span>
            </div>
            <div className="list__item-description">
                <p className="list__item-price">{price} грн.</p>
                <Button type="button" children="В корзину" className="button" onClick={() => {isOpenModal(); addToCard()}} />
            </div>
        </div>
    );
};

export default GameList;