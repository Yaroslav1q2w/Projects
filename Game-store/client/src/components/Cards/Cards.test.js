import {render, screen} from '@testing-library/react';
import '@testing-library/jest-dom'
import GameCard from "./components/GameCard";
import {Provider} from "react-redux";
import store from "../../store/store"
import userEvent from "@testing-library/user-event";
import Button from "../Button";
import GameList from "./components/GameList";


describe("test GameCard", () => {

    test('render GameCard' , () => {
        const {container} = render(<Provider store={store}><GameCard cardProps={true} /></Provider>)

        expect(container).toBeInTheDocument()
        expect(container).toMatchSnapshot()
    })

    test('render button in GameCard' , () => {
        render(<Provider store={store}><GameCard cardProps={true} /></Provider>)

        const button = screen.getByText(/add game/i)
        expect(button).toBeInTheDocument()
    })

    test('add button in GameCard' , () => {
        const handlerButton = jest.fn()
        render(<Button onClick={handlerButton}>Add game</Button>)
        userEvent.click(screen.getByText(/add game/i))
        expect(handlerButton).toHaveBeenCalledTimes(1)
    })
})


describe("test GameList", () => {

    test('render GameList' , () => {
        const {container} = render(<Provider store={store}><GameList cardProps={true} /></Provider>)

        expect(container).toBeInTheDocument()
        expect(container).toMatchSnapshot()
    })

    test('add button in GameList' , () => {
        render(<Provider store={store}><GameList cardProps={true} /></Provider>)

        const button = screen.getByText(/add game/i)
        expect(button).toBeInTheDocument()
    })

    test('add button in GameList' , () => {
        const handlerButton = jest.fn()
        render(<Button onClick={handlerButton}>Add game</Button>)
        userEvent.click(screen.getByText(/add game/i))
        expect(handlerButton).toHaveBeenCalledTimes(1)
    })
})