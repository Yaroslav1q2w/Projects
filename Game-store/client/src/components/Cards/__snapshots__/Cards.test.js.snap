// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test GameCard render GameCard 1`] = `
<div>
  <div
    class="game__item"
  >
    <img
      height="180"
      width="320"
    />
    <div
      class="game__item-details"
    >
      <div
        class="game__item-header"
      >
        <h3
          class="game__item-title"
        />
        <div
          class="star-icon"
        >
          <svg
            fill="currentColor"
            font-size="26"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3zM664.8 561.6l36.1 210.3L512 672.7 323.1 772l36.1-210.3-152.8-149L417.6 382 512 190.7 606.4 382l211.2 30.7-152.8 148.9z"
            />
          </svg>
        </div>
      </div>
      <span
        class="game__item-article"
      >
        Артикул: 
      </span>
      <div
        class="game__item-description"
      >
        <p
          class="game__item-price"
        >
           грн.
        </p>
        <button
          class="button"
          type="button"
        >
          Add game
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`test GameList render GameList 1`] = `
<div>
  <div
    class="list__item"
  >
    <img
      height="150"
      width="260"
    />
    <div
      class="list__item-details"
    >
      <div
        class="list__item-header"
      >
        <h3
          class="list__item-title"
        />
        <div
          class="list__item-favorite"
        >
          <svg
            fill="currentColor"
            font-size="30"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3zM664.8 561.6l36.1 210.3L512 672.7 323.1 772l36.1-210.3-152.8-149L417.6 382 512 190.7 606.4 382l211.2 30.7-152.8 148.9z"
            />
          </svg>
        </div>
      </div>
      <span
        class="list__item-article"
      >
        Артикул: 
      </span>
    </div>
    <div
      class="list__item-description"
    >
      <p
        class="list__item-price"
      >
         грн.
      </p>
      <button
        class="button"
        type="button"
      >
        Add game
      </button>
    </div>
  </div>
</div>
`;
